src/main/scala/io/iohk/ethereum/network/PeerActor.scala

```diff
@@ -207,21 +209,44 @@ class PeerActor(nodeInfo: NodeInfo) extends Actor with ActorLogging {
   }
 
   def handlePingMsg(rlpxConnection: RLPxConnection): Receive = {
-    case MessageReceived(ping: Ping) => rlpxConnection.sendMessage(Pong())
+    case RLPxConnectionHandler.MessageReceived(ping: Ping) => rlpxConnection.sendMessage(Pong())
   }
 
   def handleDisconnectMsg : Receive = {
-    case MessageReceived(d: Disconnect) =>
+    case RLPxConnectionHandler.MessageReceived(d: Disconnect) =>
       log.info("Received {}. Closing connection", d)
       context stop self
   }
 
+  def handleSubscriptions: Receive = {
+    case Subscribe(messageCodes) =>
+      val (senderSubscriptions, remainingSubscriptions) = messageSubscribers.partition(_.ref == sender())
+      val allMessageCodes = senderSubscriptions.flatMap(_.messageCodes).toSet ++ messageCodes
+      messageSubscribers = remainingSubscriptions :+ Subscriber(sender(), allMessageCodes)
+
+    case Unsubscribe =>
+      messageSubscribers = messageSubscribers.filterNot(_.ref == sender())
+  }
+
   class HandshakedHandler(rlpxConnection: RLPxConnection) {
 
-    def receive: Receive = handleTerminated(rlpxConnection) orElse {
-      case RLPxConnectionHandler.MessageReceived(message) => processMessage(message)
-      case s: SendMessage[_] => rlpxConnection.sendMessage(s.message)(s.enc)
-      case GetStatus => sender() ! StatusResponse(Handshaked)
+    def receive: Receive =
+      handleSubscriptions orElse handleTerminated(rlpxConnection) orElse {
+      case RLPxConnectionHandler.MessageReceived(message) =>
+        notifySubscribers(message)
+        processMessage(message)
+
+      case s: SendMessage[_] =>
+        rlpxConnection.sendMessage(s.message)(s.enc)
+
+      case GetStatus =>
+        sender() ! StatusResponse(Handshaked)
+    }
+
+    def notifySubscribers(message: Message): Unit = {
+      val subscribers = messageSubscribers.filter(_.messageCodes.contains(message.code))
+      val toNotify = subscribers.map(_.ref).toSet
+      toNotify.foreach { _ ! MessageReceived(message) }
```

Comment:
Do subscribers ever need to reply to the Message I wonder...? 
If we change this to 'forward' they can reply assuming the sender is the connection handler. 
I can't think of any scenario that requires reply yet...